### PR TITLE
Add filter named `replace` to substitute one substring with another

### PR DESCRIPTION
**Pull request by sibljon**
According to the contribution guideline, I was supposed to open an Issue to discuss this first, but since this is something I need in my own project and I needed it now, it seemed simplest to just discuss in this Pull Request. That being said, feel free to not accept this, of course! I'm also happy to make changes, as desired.

### CODE REVIEW

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift [hunk 1]
```diff
@@ -30,6 +30,13 @@ final class StencilTemplate: Stencil.Template, Template {
     private static func sourceryEnvironment() -> Stencil.Environment {
         let ext = Stencil.Extension()
         ext.registerFilter("upperFirst", filter: Filter<String>.make({ $0.upperFirst() }))
+        ext.registerFilterWithTwoArguments("replace", filter: { (s1: Any?, s2: Any, s3: Any) -> Any? in
+            if let s1 = s1 as? String, let s2 = s2 as? String, let s3 = s3 as? String {
+                return s1.replacingOccurrences(of: s2, with: s3)
+            }
+            return s1
```
Comment (sibljon) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: I wasn't sure what's the best way to fail if the arguments are of the wrong type (not Strings), so I defaulted to just passing through the value.

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift [hunk 2]
```diff
@@ -85,6 +92,15 @@ extension Annotated {
...
 extension Stencil.Extension {
+    
+    func registerFilterWithTwoArguments<A, B>(_ name: String, filter: @escaping (Any?, A, B) throws -> Any?) {
+        registerFilter(name) { (any, args) throws -> Any? in
+            guard args.count == 3, let argA = args[0] as? A, let argB = args[2] as? B else {
+                throw TemplateSyntaxError("'\(name)' filter takes two arguments: \(A.self) and \(B.self)")
+            }
+            return try filter(any, argA, argB)
+        }
+    }
```
Comment (sibljon) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: I'm also happy to make changes, as desired. I tried to follow suit by wrapping this in a `registerStringFilterWithTwoArguments`, but I couldn't quite get Swift compiler errors, so I left it as-is. I also wasn't sure what's the best way to fail

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift (hunk 1 above)
Comment (ilyapuchka) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: Basically we should throw error if parameters do not validate. Like in `registerFilterWithArguments`.

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift (hunk 2 above)
Comment (krzysztofzablocki) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: You can simplify your code:
```swift
func registerFilterWithTwoArguments<T, A, B>(_ name: String, filter: @escaping (T, A, B) throws -> Any?) {
        registerFilter(name) { (any, args) throws -> Any? in
            guard let type = any as? T else { return any } 
            guard args.count == 3, let argA = args[0] as? A, let argB = args[2] as? B else {
                throw TemplateSyntaxError("'\(name)' filter takes two arguments: \(A.self) and \(B.self)")
            }
            return try filter(type, argA, argB)
        }
   }
```

Then at your call site you can get rid of any guess work
```swift
 ext.registerFilterWithTwoArguments("replace", filter: { (source: String, substring: String, replacement: String) -> Any? in
              return source.replacingOccurrences(of: substring, with: replacement)
         })
```

Comment (ilyapuchka) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: Can we reuse `registerFilterWithArguments` for that, using tuple as arguments type?

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift (hunk 1 above)
Comment (krzysztofzablocki) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: but only parameters, we don't throw on the value type (and careful with optionality)

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift (hunk 2 above)
Comment (krzysztofzablocki) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: if the cast will work that's even better 👍

**Comment by sibljon** [comment]
> Thanks for contribution @sibljon !
> I think we might also add filters like appendPrefix/appendSuffix on top of that. What do you think @krzysztofzablocki

My pleasure!

As a matter of fact, that's also a use case of mine. Is there any advantage to having appendPrefix/appendSuffix as filters vs. `{{ foo }}SomeSuffix` ?

Also, it seems adding the ability to preserve camel casing would be important, though I'm not exactly sure how that would work. Is there a way to compound two filters together currently, or rather, to use the result of one filter as an input to another?

**Comment by ilyapuchka** [comment]
@sibljon ah, nevermind about "appendPrefix"/"appendSuffix", it was stupid, just having it like you pointed is more than enough 😅

Speaking about combining filters yes, you can combine them, we do that for example when combining `implements`/`based`/etc with `count` filter

File: Sourcery/Generating/Template/Stencil/StencilTemplate.swift (hunk 2 above)
Comment (sibljon) [inline Sourcery/Generating/Template/Stencil/StencilTemplate.swift]: Good idea @krzysztofzablocki I just amended my commit to 

> Can we reuse registerFilterWithArguments for that, using tuple as arguments type?

Also a good idea, @ilyapuchka. However, I just gave this a shot and came up short :-/

**Comment by krzysztofzablocki** [comment]
thanks for contributing 👍

**Pull request merged**